ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -0,0 +1,250 @@
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public void put(MessageWriteSet messageSetToWrite) throws StoreException {
+    checkStarted();
+    if (messageSetToWrite.getMessageSetInfo().isEmpty()) {
+      throw new IllegalArgumentException("Message write set cannot be empty");
+    }
+    checkDuplicates(messageSetToWrite);
+
+    // Write the blobs in the message set
+    CloudWriteChannel cloudWriter = new CloudWriteChannel(this, messageSetToWrite.getMessageSetInfo());
+    try {
+      messageSetToWrite.writeTo(cloudWriter);
+    } catch (IOException ex) {
+      throw new StoreException(ex, StoreErrorCodes.IOError);
+    }
+  }
+
+  /**
+   * Upload the blob to the cloud destination.
+   * @param messageInfo the {@link MessageInfo} containing blob metadata.
+   * @param messageBuf the bytes to be uploaded.
+   * @param size the number of bytes to upload.
+   * @throws Exception
+   */
+  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size) throws Exception {
+    boolean performUpload = messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time && !messageInfo.isDeleted();
+    // May need to encrypt buffer before upload
+    if (performUpload) {
+      BlobId blobId = (BlobId) messageInfo.getStoreKey();
+      cloudDestination.uploadBlob(blobId, size, new ByteBufferInputStream(messageBuf));
+    }
+  }
+
+  @Override
+  public void delete(MessageWriteSet messageSetToDelete) throws StoreException {
+    checkStarted();
+    checkDuplicates(messageSetToDelete);
+
+    try {
+      for (MessageInfo msgInfo : messageSetToDelete.getMessageSetInfo()) {
+        BlobId blobId = (BlobId) msgInfo.getStoreKey();
+        cloudDestination.deleteBlob(blobId);
+      }
+    } catch (Exception ex) {
+      throw new StoreException(ex, StoreErrorCodes.IOError);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * Currently, the only supported operation is to set the TTL to infinite (i.e. no arbitrary increase or decrease)
+   * @param messageSetToUpdate The list of messages that need to be updated
+   * @throws StoreException if there is a problem persisting the operation in the store.
+   */
+  @Override
+  public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException {
+    checkStarted();
+    // This is currently a no-op because we skipped uploading the blob on receipt of the PUT record since it had a TTL.
+  }
+
+  @Override
+  public FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws StoreException {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public Set<StoreKey> findMissingKeys(List<StoreKey> keys) throws StoreException {
+    checkStarted();
+    // Check existence of keys in cloud container
+    Set<StoreKey> missingKeys = new HashSet<>();
+    try {
+      for (StoreKey key : keys) {
+        BlobId blobId = (BlobId) key;
+        if (!cloudDestination.doesBlobExist(blobId)) {
```

Comment:
how costly is this? This will happen very often

Reply:
Currently it will be expensive because needs to call Azure to check at the blob level.  We are planning to add an index (Azure Table perhaps) to allow batch lookups.  When that is ready, I will add a batch existence check method to CloudDestination and call it from here.